[OpenMP 6.0] Parse/Sema support for reduction over private variable with reduction clause.

chandraghale opened the pull request:
Initial Parse/Sema support for reduction over private variable with reduction clause.
Section 7.6.10 in  in OpenMP 6.0 spec. 
-  list item in a reduction clause can now be private in the enclosing context.
-  Added support for _original-sharing-modifier_  with reduction clause.

Commit:
Parse/Sema support for reduction over private variable with reduction clause

Commit:
formating fix

alexey-bataev commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -4668,6 +4668,34 @@ bool Parser::ParseOpenMPVarList(OpenMPDirectiveKind DKind,
       assert(Tok.is(tok::comma) && "Expected comma.");
       (void)ConsumeToken();
     }
+    // Handle original(private / shared) Modifier
+    if (Kind == OMPC_reduction && getLangOpts().OpenMP >= 60 &&
+        Tok.is(tok::identifier) && PP.getSpelling(Tok) == "original" &&
+        NextToken().is(tok::l_paren)) {
+      // Parse original(private) modifier.
+      ConsumeToken();
+      BalancedDelimiterTracker ParenT(*this, tok::l_paren, tok::r_paren);
+      ParenT.consumeOpen();
+      if (Tok.is(tok::kw_private)) {
+        Data.OriginalSharingModifier = OMPC_ORIGINAL_SHARING_private;
+        Data.OriginalSharingModifierLoc = Tok.getLocation();
+        ConsumeToken();
+      } else if (Tok.is(tok::identifier) &&
+                 (PP.getSpelling(Tok) == "shared" ||
+                  PP.getSpelling(Tok) == "default")) {
+        Data.OriginalSharingModifier = OMPC_ORIGINAL_SHARING_shared;
+        Data.OriginalSharingModifierLoc = Tok.getLocation();
+        ConsumeToken();
+      } else {
+        Diag(Tok.getLocation(), diag::err_expected)
+            << "'private or shared or default'";
+        SkipUntil(tok::r_paren);
+        return false;
+      }
+      ParenT.consumeClose();
+      assert(Tok.is(tok::comma) && "Expected comma.");
+      (void)ConsumeToken();
+    }
```
Why do you need this? Just a proper check in DSAStack should be enough

chandraghale replied:
@alexey-bataev  "I'm sorry, I didn't understand your comment. The changes are made to consume the newly added original-sharing-modifier [OpenMP 6.0 7.6.10]  For example:
```
#pragma omp parallel reduction(original(private), +: x)
#pragma omp parallel reduction(original(shared), *: y)
```
Isn't it necessary for the parser to recognize the original(...) syntax and record the modifier? I suppose DSAStack does not natively handle this modifier."

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -3678,6 +3678,9 @@ class OMPReductionClause final
   DeclarationNameInfo NameInfo;
 
+  /// Private variable reduction flag
+  llvm::SmallVector<bool, 8> IsPrivateVarReductionFlags;
```
No SmallVector is allowed here, use tail-allocated storage for extra members (`llvm::TrailingObjects`)

alexey-bataev commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -4668,6 +4668,34 @@ bool Parser::ParseOpenMPVarList(OpenMPDirectiveKind DKind,
+        return false;
+      }
+      ParenT.consumeClose();
+      assert(Tok.is(tok::comma) && "Expected comma.");
```
This should be an error, not an assertion, otherwise compiler will crash if the token is not a comma

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -17391,7 +17392,8 @@ OMPClause *SemaOpenMP::ActOnOpenMPVarListClause(OpenMPClauseKind Kind,
         VarList, static_cast<OpenMPReductionClauseModifier>(ExtraModifier),
         StartLoc, LParenLoc, ExtraModifierLoc, ColonLoc, EndLoc,
-        Data.ReductionOrMapperIdScopeSpec, Data.ReductionOrMapperId);
+        Data.ReductionOrMapperIdScopeSpec, Data.ReductionOrMapperId, {},
+        static_cast<OpenMPOriginalSharingModifier>(OriginalSharingModifier));
```
Better to pack these members into struct already, there are too many parameters in the functions

alexey-bataev commented on clang/test/OpenMP/parallel_for_reduction_messages.cpp:
```diff
@@ -226,7 +228,12 @@ T tmain(T argc) {
-
+#if defined(_OPENMP) && (_OPENMP >= 202411)
+#pragma omp parallel private(fl)
+#pragma omp for reduction(original(abcxx),+:fl) // expected-error {{private or shared or default}} expected-warning {{extra tokens at the end of '#pragma omp for'}} 
```
More tests are required, including incorrect syntax, templates, etc.

Commit:
Addressed review comments and made necessary modifications

chandraghale commented:
1. Used tail-allocated storage for extra members added in OMPreductionClause. 
2. Modified assertion to diagnostic in parsing logic. 
3. Reduction clause modifiers reduction-modifier and original-sharingmodifier packed into existing structure. 
4. Added additional lit test cases.

Commit:
Addressed Review comments

chandraghale commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -11721,6 +11721,12 @@ void OMPClauseReader::VisitOMPReductionClause(OMPReductionClause *C) {
+  unsigned NumFlags = Record.readInt();
+  SmallVector<bool, 16> Flags;
+  Flags.reserve(NumFlags);
+  for (unsigned i = 0; i != NumFlags; ++i)
```
Done

Commit:
Formatting fix

chandraghale commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -18933,12 +18945,35 @@ static bool actOnOMPReductionKindClause(
+           RD.OrigSharingModifier != OMPC_ORIGINAL_SHARING_shared);
+      // Disable private handling for OpenMP versions <= 5.2
+      if (S.getLangOpts().OpenMP <= 52)
+        IsPrivate = false;
```
Modified, now IsPrivate flag check is enclosed for OpenMP 5.2 and later. For versions ≤5.2, the DSA private check falls back to the existing code.

chandraghale commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -18933,12 +18945,35 @@ static bool actOnOMPReductionKindClause(
+      // Check if the construct is orphaned (has no enclosing OpenMP context)
+      IsOrphaned = (ParentDir == OMPD_unknown);
+      IsPrivate =
+          ((isOpenMPPrivate(DVar.CKind) && DVar.CKind != OMPC_reduction &&
+            isOpenMPWorksharingDirective(CurrDir) &&
+            !isOpenMPParallelDirective(CurrDir) &&
+            !isOpenMPTeamsDirective(CurrDir) &&
+            !isOpenMPSimdDirective(ParentDir)) ||
+           (IsOrphaned && DVar.CKind == OMPC_unknown) ||
+           RD.OrigSharingModifier != OMPC_ORIGINAL_SHARING_shared);
```
done

chandraghale commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -18933,12 +18945,35 @@ static bool actOnOMPReductionKindClause(
+      OpenMPDirectiveKind CurrDir = Stack->getCurrentDirective();
+      OpenMPDirectiveKind ParentDir = Stack->getParentDirective();
+      // Check if the construct is orphaned (has no enclosing OpenMP context)
+      IsOrphaned = (ParentDir == OMPD_unknown);
```
done

chandraghale commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -3757,6 +3768,31 @@ class OMPReductionClause final
+  void setPrivateVariableReductionFlags(ArrayRef<bool> Flags) {
+    assert(Flags.size() == varlist_size() &&
+           "Number of private flags does not match vars");
+    std::copy(Flags.begin(), Flags.end(), getTrailingObjects<bool>());
```
done

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -3757,6 +3768,31 @@ class OMPReductionClause final
+
+  /// Get the list of help private variable reduction flags
+  MutableArrayRef<bool> getPrivateVariableReductionFlags() {
+    return MutableArrayRef<bool>(getTrailingObjects<bool>(), varlist_size());
```
```suggestion
    return MutableArrayRef(getTrailingObjects<bool>(), varlist_size());
```
?

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -3757,6 +3768,31 @@ class OMPReductionClause final
+    return MutableArrayRef<bool>(getTrailingObjects<bool>(), varlist_size());
+  }
+  ArrayRef<bool> getPrivateVariableReductionFlags() const {
+    return ArrayRef<bool>(getTrailingObjects<bool>(), varlist_size());
```
```suggestion
    return ArrayRef(getTrailingObjects<bool>(), varlist_size());
```

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -3854,6 +3890,7 @@ class OMPReductionClause final
   /// region with this clause.
   /// \param PostUpdate Expression that must be executed after exit from the
   /// OpenMP region with this clause.
+  /// \pram IsPrivateVarReduction array for private variable reduction flags
```
```suggestion
  /// \param IsPrivateVarReduction array for private variable reduction flags
```

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1160,6 +1162,12 @@ class SemaOpenMP : public SemaBase {
+    struct OpenMPReductionClauseModifiers {
+      int ExtraModifier;
+      int OriginalSharingModifier;
+      OpenMPReductionClauseModifiers(int extra, int original)
```
```suggestion
      OpenMPReductionClauseModifiers(int Extra, int Original)
```

Commit:
Additional review comment address

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -18933,12 +18945,34 @@ static bool actOnOMPReductionKindClause(
+      IsOrphaned = ParentDir == OMPD_unknown;
+      // OpenMP 6.0: Private DSA check
+      IsPrivate =
+          (S.getLangOpts().OpenMP > 52) &&
```
Extra parens

alexey-bataev reviewed (approved):
LG with a nit

chandraghale commented:
Thank you for the review.
